crates/uv/src/commands/pip/compile.rs

```diff
@@ -289,7 +291,7 @@ pub(crate) async fn pip_compile(
         .connectivity(connectivity)
         .index_urls(index_locations.index_urls())
         .index_strategy(index_strategy)
-        .keyring(keyring_provider)
+        .middleware_stack(MiddlewareStack::new(3, keyring_provider))
```

Comment:
I don't love this new API.



Hey @tdejager sorry I've been slow to get to this.

Could we instead:

- Only create the retry middleware if retries is non-zero
- Replace the `Keyring` option with `Option<AuthMiddleware>`
- Add a `extra_middleware: Vec<Arc<dyn Middleware>>` option

I think this should get you the same functionality, right?

Reply:
Sounds like a good improvement to me, I can also make the changes early next week if you are busy @konstin :)

Reply:
Also @zanieb and @konstin thanks again for taking a look at:)

Reply:
I'll be out next week fyi.

I missed that @konstin introduced the `MiddlewareStack` — let's make sure he's onboard with my proposal :)

Reply:
Yeah maybe having the AuthMiddleware in the stack is enough. I thought you might want to be more explicit about enabling/disabling the auth middleware. Let see what konsti thinks :)

Reply:
I think we can just have the `AuthMiddleware` populated by default and ya'll can pass `None` to explicitly remove it.

Reply:
I changed the API so it's now `.with` methods and checks for > 0 retries, but if we don't take a keyring it becomes even more verbose because we have a number of callsites, and also `KeyringProviderType` lives in `uv-configuration` so it can't be `Into<AuthMiddleware>`.

![image](https://github.com/astral-sh/uv/assets/6826232/b88c3a77-3245-4a72-b5d8-00cb42f0f837)